### PR TITLE
DAOS-10607 rebuild: put rebuild obj cache by tgt id

### PR DESCRIPTION
**Pull request by wangdi1**
Store scan objects target id, so it does not need excessive
iteration when sending the objects to the new target.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

**Comment by rpadma2** [comment]
@wangdi1 : Please add the following commit message to the PR:

Test-tag: pr offline_drain_after_snapshot offline_reintegration_with_rf

**Comment by wangdi1** [comment]
> @wangdi1 : Please add the following commit message to the PR:
> 
> Test-tag: pr offline_drain_after_snapshot offline_reintegration_with_rf

Ah, sure thanks @rpadma2

File: src/rebuild/scan.c [hunk 1]
```diff
@@ -647,10 +716,11 @@ rebuild_obj_scan_cb(daos_handle_t ch, vos_iter_entry_t *ent,
 	if (map != NULL)
 		pl_map_decref(map);
 
-	if (--arg->yield_freq == 0) {
+	if (--arg->yield_freq == 0 || --arg->obj_yield_cnt == 0) {
```
Comment (liuxuezhao) [inline src/rebuild/scan.c:719]: why need obj_yield_cnt besides yield_freq? and seems obj_yield_cnt possible become < 0 here as above for rebuild_nr do obj_yield_cnt--?

File: src/dtx/dtx_resync.c [hunk 2]
```diff
@@ -518,7 +518,7 @@ dtx_iter_cb(uuid_t co_uuid, vos_iter_entry_t *ent, void *args)
 	} else {
 		/* For pool map refresh. */
 		/* Leader: do nothing. */
-		if ((ent->ie_dtx_flags & DTE_LEADER) && dra->resync_version != dra->discard_version)
+		if ((ent->ie_dtx_flags & DTE_LEADER) && dra->discard_version == 0)
```
Comment (liuxuezhao) [inline src/dtx/dtx_resync.c:521]: may refer @Nasf-Fan 's comment of PR9744.
and this PR with quite some same code change as PR9744, won't they conflict?

**Review by liuxuezhao — APPROVED** [review]

File: src/rebuild/scan.c (hunk 1 above)
Comment (liuxuezhao) [inline src/rebuild/scan.c:719]: why need obj_yield_cnt besides yield_freq? and seems obj_yield_cnt possible become < 0 here as above for rebuild_nr do obj_yield_cnt--?

**Review by liuxuezhao — APPROVED** [review]

File: src/rebuild/scan.c (hunk 1 above)
Comment (wangdi1) [reply src/rebuild/scan.c:719]: there are 2 cases needs to be yield.

1. yield_freq is increased to 4096, if it reaches to 0, then it means enough scan, even if there no objects.
2. If there are already enough objects needs to be sent, then yield to send.

**Review by liuxuezhao — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**